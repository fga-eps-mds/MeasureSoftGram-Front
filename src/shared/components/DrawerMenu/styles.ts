import styled from 'styled-components';

export const ScrollDiv = styled.div`
  width: auto;
  flexdirection: column;
  overflow-y: auto;
  margin-top: 50px;
`;
