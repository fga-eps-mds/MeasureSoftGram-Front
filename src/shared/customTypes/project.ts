import { ButtonProps } from '@mui/material';

export interface CollectedMetric {
  id: number;
  measure_id: number;
  value: number;
  created_at: string;
}

export interface MeasuresHistoryResult {
  id: number;
  key: string;
  name: string;
  description: string;
  history: Array<CollectedMetric>;
}

export interface MeasuresHistory {
  count: string;
  next: string;
  previous: string;
  results: Array<MeasuresHistoryResult>;
}

export interface Project {
  id: number;
  name: string;
  description: string;
  github_url: string;
  created_at: string;
  updated_at: string;
}

interface DefaultAttr {
  id: number;
  key: string;
  name: string;
  description: string | null;
}

interface PreConfigSubCharacteristics extends DefaultAttr{
  measures: Array<DefaultAttr>
}

export interface PreConfigEntitiesRelationship extends DefaultAttr {
  subcharacteristics: Array<PreConfigSubCharacteristics>
}

export interface CharacteristicWithBalanceMatrix {
  id: number;
  key: string;
  value: number;
  correlations: {
    '+': string[];
    '-': string[]
  }
}

export interface ValuesCommitted {
  [key: string]: number
}

export interface Changes {
  characteristic_key: keyof ValuesCommitted;
  delta: number
}

export interface ReleaseGoal {
  release_name: string;
  start_at: string;
  end_at: string;
  changes: Changes[];
}

export interface ButtonType extends Omit<Partial<ButtonProps>, 'color'> {
  label: string;
  onClick: () => void;
  backgroundColor: string;
  color: string;
  variant?: ButtonProps['variant'];
}

interface SqcValue {
  id: number;
  value: number;
  created_at: string;
}

interface RepositoriesSqcHistoryResult {
  id: number;
  url: string;
  name: string;
  history: Array<SqcValue>
}

export interface RepositoriesSqcHistory {
  count: number;
  results: Array<RepositoriesSqcHistoryResult>
}
