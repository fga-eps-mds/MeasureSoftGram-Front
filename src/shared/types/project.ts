interface CollectedMetric {
  id: number;
  measure_id: number;
  value: number;
  created_at: string;
}

export interface MeasuresHistoryResult {
  id: number,
  key: string,
  name: string,
  description: string,
  history: Array<CollectedMetric>
}

export interface MeasuresHistory {
  count: string;
  next: string;
	previous: string;
	results: Array<MeasuresHistoryResult>
}

export interface Project {
  id: number;
  name: string;
  description: string;
  github_url: string;
  created_at: string;
  updated_at: string;
}
