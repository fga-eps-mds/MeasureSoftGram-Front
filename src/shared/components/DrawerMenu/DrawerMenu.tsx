import * as React from 'react';
import Box from '@mui/material/Box';
import Drawer from '@mui/material/Drawer';
import { Button, Typography } from '@mui/material';
import { ButtonType } from '@customTypes/project';
import * as Styles from './styles';

const { ScrollDiv } = Styles;
interface DrawerMenuProps {
  children: JSX.Element;
  open: boolean;
  buttons?: Array<ButtonType>;
  subtitle?: string;
  title?: string;
}

const DrawerMenu = ({ children, open, buttons, title, subtitle }: DrawerMenuProps) => {
  const renderButtons = () => {
    if (buttons)
      return (
        <Box marginTop="16px">
          {buttons.map((button) => (
            <Button
              key={Math.random()}
              variant={button.variant}
              sx={{
                backgroundColor: button.backgroundColor,
                color: button.color,
                padding: '9px 20px',
                marginRight: '16px'
              }}
              onClick={button.onClick}
            >
              {button.label}
            </Button>
          ))}
        </Box>
      );
  };

  const renderTitle = () => {
    if (title)
      return (
        <Box>
          <Typography variant="h3">{title}</Typography>
          {!!subtitle && <Typography mt="12px">{subtitle}</Typography>}
        </Box>
      );
  };

  return (
    <Drawer anchor="right" open={open}>
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="space-between"
        sx={{ padding: '46px 36px 34px', height: '100%' }}
      >
        <ScrollDiv>
          {renderTitle()}
          {children}
        </ScrollDiv>
        {renderButtons()}
      </Box>
    </Drawer>
  );
};

export default DrawerMenu;
