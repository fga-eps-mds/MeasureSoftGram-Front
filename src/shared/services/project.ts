/* eslint-disable class-methods-use-this */
import { PreConfigEntitiesRelationship, MeasuresHistory, ReleaseGoal, RepositoriesSqcHistory } from '@customTypes/project';
import { Data } from '@customTypes/preConfig';
import api from './api';

class ProjectQuery {
  async getProjectById(organizationId: string, id: string) {
    return api.get(`/organizations/${organizationId}/products/${id}/`);
  }

  async getAllProjects(organizationId: string) {
    return api.get(`/organizations/${organizationId}/repository/`);
  }

  async getProjectMeasuresHistory(organizationId: string, projectId: string) {
    const url = `organizations/${organizationId}/repository/${projectId}/history/measures/`;
    return api.get<MeasuresHistory>(url);
  }

  async getPreConfig(organizationId: string, productId: string) {
    return api.get(`/organizations/${organizationId}/products/${productId}/current/pre-config/`);
  }

  postPreConfig(organizationId: string, productId: string, data: { name: string; data: Data }) {
    api.post(`/organizations/${organizationId}/products/${productId}/create/pre-config/`, data);
  }

  async getPreConfigEntitiesRelationship(organizationId: string, projectId: string) {
    const url = `organizations/${organizationId}/products/${projectId}/entity-relationship-tree/`;
    return api.get<Array<PreConfigEntitiesRelationship>>(url);
  }

  async createProjectReleaseGoal(organizationId: string, projectId: string, data: ReleaseGoal) {
    const url = `organizations/${organizationId}/products/${projectId}/create/goal/`;
    return api.post(url, data);
  }

  async getProductRepositoriesSqcHistory(organizationId: string, productId: string) {
    const url = `organizations/${organizationId}/products/${productId}/repositories-sqc-historical-values/`;
    return api.get<RepositoriesSqcHistory>(url);
  }
}

export const projectQuery = new ProjectQuery();
Object.freeze(projectQuery);
