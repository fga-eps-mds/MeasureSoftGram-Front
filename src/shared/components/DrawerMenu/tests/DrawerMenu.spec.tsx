import '@testing-library/jest-dom';

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ButtonType } from '@customTypes/project';

import DrawerMenu from '../DrawerMenu';

const BUTTON_NAME = 'ButtonLabel';

const BUTTONS: Array<ButtonType> = [
  {
    label: BUTTON_NAME,
    onClick: jest.fn(),
    backgroundColor: '#FFF',
    color: '#000',
    variant: 'outlined'
  }
];

describe('<DrawerMenu />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <DrawerMenu open>
          <div />
        </DrawerMenu>
      );
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao Snapshot com botões', () => {
      const tree = render(
        <DrawerMenu open buttons={BUTTONS}>
          <div />
        </DrawerMenu>
      );
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao Snapshot com componente fechado', () => {
      const tree = render(
        <DrawerMenu open={false}>
          <div />
        </DrawerMenu>
      );
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Comportamento', () => {
    it('Botão dentro do DrawerMenu deve ser clicavel', () => {
      const { getByText } = render(
        <DrawerMenu open buttons={BUTTONS}>
          <div />
        </DrawerMenu>
      );

      const button = getByText(BUTTON_NAME);

      fireEvent.click(button);

      expect(BUTTONS[0].onClick).toBeCalled();
    });
  });
});
