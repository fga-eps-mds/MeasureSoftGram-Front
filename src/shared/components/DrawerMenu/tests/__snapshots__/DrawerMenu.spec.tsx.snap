// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrawerMenu /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDrawer-root MuiDrawer-modal MuiModal-root css-1u2w381-MuiModal-root-MuiDrawer-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorRight css-1160xiw-MuiPaper-root-MuiDrawer-paper"
        tabindex="-1"
      >
        <div
          class="MuiBox-root css-1m45ylu"
        >
          <div
            class="styles__ScrollDiv-sc-1casz44-0 eTyGvR"
          >
            <div />
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<DrawerMenu /> Snapshot Deve corresponder ao Snapshot com bot√µes 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiDrawer-root MuiDrawer-modal MuiModal-root css-1u2w381-MuiModal-root-MuiDrawer-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-testid="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation16 MuiDrawer-paper MuiDrawer-paperAnchorRight css-1160xiw-MuiPaper-root-MuiDrawer-paper"
        tabindex="-1"
      >
        <div
          class="MuiBox-root css-1m45ylu"
        >
          <div
            class="styles__ScrollDiv-sc-1casz44-0 eTyGvR"
          >
            <div />
          </div>
          <div
            class="MuiBox-root css-1yuhvjn"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-iss8vz-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              ButtonLabel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        data-testid="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<DrawerMenu /> Snapshot Deve corresponder ao Snapshot com componente fechado 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style=""
  >
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
